src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -956,13 +957,46 @@ public bool StartsWith(string value)
             return StartsWith(value, StringComparison.CurrentCulture);
         }
 
+        // A hint for the inliner to always prescan this method
+        // and take the constant input path
+        [Intrinsic]
         public bool StartsWith(string value, StringComparison comparisonType)
         {
             if (value is null)
             {
                 throw new ArgumentNullException(nameof(value));
             }
 
+#if TARGET_64BIT && !MONO && !BIGENDIAN
+            if (RuntimeHelpers.IsKnownConstant(value) && RuntimeHelpers.IsKnownConstant((int)comparisonType) &&
```

Comment:
What is the behavior when somebody specifies constant `comparisonType` and non-constant `value`? Is the method going to be aggressively inlined in that case as well?

Reply:
That's a good point 🤔 I guess the only solution for this problem is to replace the 2nd `RuntimeHelpers.IsKnownConstant((int)comparisonType)` with `RuntimeHelpers.IsKnownTrue(comparisonType == Ordinal)`

Reply:
Something like this:
``
bool addMultiplier = (count(IsKnownTrue for const arg) + 
                      count(IsKnowConstant for const arg) - 
                      count(IsKnownTrue for non const arg) - 
                      count(IsKnowConstant for non const arg)) > 0
```

Reply:
I am not sure whether it would solve the problem. There are still two conditions even with this change. The aggressive inlining should kick in only when both of the conditions are satisfied.

Reply:
@jkotas how about moving everything to a non-inlineable fallback and mark StartsWith with AggressiveInlining so non-constant case won't regress and won't be inlined

Reply:
I think that would work. I assume you won't need to mark it as intrinsic with that approach. It has a risk of significantly increasing JIT-time if used broadly.